packages/dds/tree/src/shared-tree-core/branch.ts

```diff
@@ -492,6 +443,86 @@ export class SharedTreeBranch<TEditor extends ChangeFamilyEditor, TChange> {
 		return rebaseResult;
 	}
 
+	/**
+	 * Remove a range of commits from this branch.
+	 * @param commit - All commits after (but not including) this commit, up to and including the branch's head, will be removed.
+	 * @returns The net change to this branch and the commits that were removed from this branch.
+	 */
+	public removeAfter(
+		commit: GraphCommit<TChange>,
+	): [change: TaggedChange<TChange> | undefined, removedCommits: GraphCommit<TChange>[]] {
+		if (commit === this.head) {
+			return [undefined, []];
+		}
+
+		const removedCommits: GraphCommit<TChange>[] = [];
+		const inverses: TaggedChange<TChange>[] = [];
+		findAncestor([this.head, removedCommits], (c) => {
+			if (c !== commit) {
+				const revision = this.mintRevisionTag();
+				const inverse = this.changeFamily.rebaser.changeRevision(
+					this.changeFamily.rebaser.invert(c, true, revision),
+					revision,
+					c.revision,
+				);
+
+				inverses.push(tagRollbackInverse(inverse, revision, c.revision));
+				return false;
+			}
```

Comment:
I'm not a fan of having side effects within the `findAncestor` callstack. I don't think the contract for `findAncestor` should be so specific as to guarantee this will work as expected (it currently hints that way but I don't think it should). Seems like the legibility of the code would go up anyway if we pulled this out.

Reply:
Hmm... I like this way. If you pull it out, then you have to:
1. Write an additional loop...
2. ...which is a backwards loop, which means you can't use `for of` but have to use indices...
3. ...which means you have to assert that the read in every iteration is not undefined...

You can see this in the equivalent code that was removed from `abortTransaction`. In terms of legibility and code niceness, I definitely prefer this version. 

Is there a way we can strengthen the contract of `findAncestor` to your liking? It already says "predicate - a function which will be evaluated on every ancestor of `descendant` until it returns true." Are you concerned that it doesn't specify the order that it evaluates the ancestors, or something? We could change it to "...on each ancestor **(in ascending order**) of descendant...".

I'm generally a fan of writing "functional-style" code - i.e. chaining steps together sequentially rather than trying to do two things at once like is being done here. But that's because it usually makes the code prettier - in this case it makes it worse.

Reply:
Well, I guess the true "functional-style" way to do it would be to `reverse()` the array before iterating. But I don't think we want to do that for perf :)